### DISCUSSION TITLE
How to use social auth in supabase via React Native Expo (now that auth.expo.io is and startAsync moving to end of life)

### DISCUSSION THREAD

**Discussion by hoochani** [supabase/supabase#13290] [Questions]
Hi, 

For react native (+ expo) there's currently a pretty workable setup to leverage supabase's social auth providers. There's an article here that does a pretty good job of explaining it:

https://dev.to/fedorish/google-sign-in-using-supabase-and-react-native-expo-14jf

And here is my sample react native sign in code for using google auth via supabase. 

```
import {
  makeRedirectUri,
  startAsync,
} from 'expo-auth-session';

...

  const signInWithGoogle = async () => {
    // This will create a redirectUri

    // This builds the URL for the signin page based on the
    // expo URL scheme specified in the config. Ensures it works
    // in local dev and in production.
    const redirectUrl = makeRedirectUri({
      path: 'signin',
    });

    // authUrl: https://{YOUR_PROJECT_REFERENCE_ID}.supabase.co
    // returnURL: the redirectUrl you created above.
    const authResponse = await startAsync({
      authUrl: `${supabaseUrl}/auth/v1/authorize?provider=google&redirect_to=${redirectUrl}`,
      returnUrl: redirectUrl,
      projectNameForProxy: '@hoochani/timetogive',
    });

    // If the user successfully signs in
    // we will have access to an accessToken and an refreshToken
    // and then we'll use setSession (https://supabase.com/docs/reference/javascript/auth-setsession)
    // to create a Supabase-session using these token
    if (authResponse.type === 'success') {
      supabase.auth.setSession({
        access_token: authResponse.params.access_token,
        refresh_token: authResponse.params.refresh_token,
      });
    }
  };
```

Super simple, and it works. 

However, by leveraging the `startAsync` method from `expo-auth-session` this approach uses an intermediate proxy / redirection service hosted at `https://auth.expo.io`. 

Due to a recent security hitch (now fixed), the general guidance is to stop using it and no doubt it will become end of life at some point soon:

<img width="543" alt="image" src="https://user-images.githubusercontent.com/124574150/228471655-6fb7fb5d-1a9b-4c6e-9bf0-ced6e4b26705.png">

The advice is, therefore, that we should "configure your app to navigate to the third-party authentication provider directly instead of using this service"

That's cool. But how to do we do that but still leverage supabase as the central auth provider?

Expo has a whole bunch of examples of doing this directly with 3rd party social auth providers (without the auth.expo.io proxy) . See here:

https://docs.expo.dev/guides/authentication/

In order to leverage these they take advantage of a react hook from Expo called `useAuthRequest`. An example, of doing this against Cognito (as an example) is found here:

https://docs.expo.dev/guides/authentication/#cognito

Having trawled around now for quite a while now, I can't find a way of configuring `useAuthRequest` to leverage supabase auth. My initial use case to leverage supabase + google social auth. But supabase doesn't seem "discoverable" like other providers. And I am not knowledgeable enough in this space (particularly) around goTrue to set the appropriate config (if there is such an approach). 

Grateful for help.

**Comment by JNGO27** [comment]
I could not figure out for the life of me. I almost gave up on supabase after ripping my hair out for this exact problem. However, I ended up doing this work around, and it works.

```js
import { makeRedirectUri } from "expo-auth-session";
import * as Linking from "expo-linking";
```

```js
  const deepLinkSupaBaseUrl = Linking.createURL(SUPABASE_URL);

  const redirectUri = makeRedirectUri({
    path: deepLinkSupaBaseUrl,
  });

  const signInWithProvider = async () => {
    const { data } = await supabaseConfig.auth.signInWithOAuth({
      provider,
      options: {
        redirectTo: redirectUri,
        queryParams: {
          access_type: "offline",
          prompt: "consent",
        },
      },
    });

    Linking.openURL(data.url as string);

    Linking.addEventListener("url", (event) => {
      const { url } = event;
      const [accessToken, refreshToken] = getTokens(url);

      supabaseConfig.auth.setSession({
        access_token: accessToken,
        refresh_token: refreshToken,
      });
    });
```

```js
export function getTokens(url: string) {
  const hashFragment = url.split("#")[1];
  const andFragment = url.split("&")[4];

  const accessToken = new URLSearchParams(hashFragment).get(
    "access_token",
  ) as string;
  const refreshToken = new URLSearchParams(andFragment).get(
    "refresh_token",
  ) as string;

  return [accessToken, refreshToken];
}
```

I don't know if this is a good or bad answer. But, it works. If you find another way. Feel free to show an alternative if you find one.

**Reply by hoochani** [reply]
Super interesting @JNGO27 - thanks a lot for this - you have seriously gone deep on this to find a working solution. I will experiment with this tomorrow as well. 

Would be fab to get some input from the Supabase team...

It would be awesome if Supabase could appear on the list of supported Expo Authentication guides - https://docs.expo.dev/guides/authentication/. Appreciate this is an Expo doc, but suspect it needs some guidance from supabase / GoTrue team on how to set it up. Happy to make PR's to the docs on Supabase and Expo once we know the way forward.
